kernel: msgq: adding support for `k_msgq_put_front` and sample code

alexpaschoaletto opened the pull request:
Following the examples of [FreeRTOS](https://www.freertos.org/Documentation/02-Kernel/04-API-references/06-Queues/07-xQueueSendToFront) and [RTEMS](https://docs.rtems.org/docs/main/c-user/message/directives.html#rtems-message-queue-urgent), this PR introduces a ` k_msgq_put_front` API for sending messages to the head of the queue, in a LIFO scheme. It also adds a simple sample code that demonstrates how it works.

Due to the fact that both versions of `k_msgq_put` share most of the code, this contribution creates an internal `put_msg_in_queue` function which inserts the message at the front or back of the queue depending on the flag `put_at_back` passed to it. Now the implementation looks like this:

```c
/* internal function */
int put_msg_in_queue(struct k_msgq *msgq, const void *data, k_timeout_t timeout, bool put_at_back)
{
        /*
            puts message to front or back of the queue
            depending on the value or put_at_back
        */
}

int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout)
{
	return put_msg_in_queue(msgq, data, timeout, true);
}

int z_impl_k_msgq_put_front(struct k_msgq *msgq, const void *data, k_timeout_t timeout)
{
	return put_msg_in_queue(msgq, data, timeout, false);
}
```

_note: some work concerning mainly docs and ci/cd still needs updating, but the bulk of the contribution is here. I'd like to have some opinions on this contribution before taking the time to do everyhing properly._ 

Closes #46339

andyross commented on kernel/msg_q.c:
```diff
@@ -187,6 +187,100 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
+
+			/* give message to waiting thread */
+			(void)memcpy(pending_thread->base.swap_data, data,
+			       msgq->msg_size);
```
Not loving the duplication of the implementation code.  Ideally all this should be shared.  In particular maintenance-quality-issues are at their most important in code surrounding memory copies into/out-of caller pointers (which can cross security boundaries if CONFIG_USERSPACE=y!).  Might take some refactoring to make that happen, obviously.

alexpaschoaletto replied:
I understand and agree, but at first I really didn't want to touch in the already existing `k_msgq_put` function. So should I do something like this, then?

```c
int _copy_to_queue ( ... , bool should_copy_to_back) {
  // shared implementation goes here
}

int z_impl_k_msgq_put( ... ) {
  return _copy_to_queue( ... , true);
}

int z_impl_k_msgq_put_urgent( ... ) {
  return _copy_to_queue( ..., , false);
}
```

alexpaschoaletto replied:
FWIW I have done just that to avoid code duplication.

andyross commented on samples/basic/msg_queue/src/main.c:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	received[BUF_SIZE - 1] = '\0';
+	/* we expect to see CBA012345... */
```
While this is a good note for understanding, it would be even better for this to be a **test** that actually checked messages were received in the expected order.

alexpaschoaletto replied:
I agree. For now I have refactored the sample code to have a regex on the tests. I intend to do the proper ZTEST thing as you suggested as well, but I'd like to settle on the implementation before that.

andyross reviewed (commented):
A design nitpick around the cut/paste feel of the change.

Also: please isolate changes to the sample to a separate commit from the semantic change in the kernel.

And while a sample **can** be used as a test case, it's really best to put an actual test case of this functionality into tests/kernel/msgq/msgq_api

TaiJuWu commented on kernel/msg_q.c:
```diff
@@ -187,6 +187,100 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
+		/* message queue isn't full */
+		pending_thread = z_unpend_first_thread(&msgq->wait_q);
+		if (unlikely(pending_thread != NULL)) {
+			SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put, msgq, timeout, 0);
```
It is good to add another trace for debug.

alexpaschoaletto replied:
I like this idea, however I'm not certain on how to refactor this part. would this be just changing `put` by `put_urgent` (or whatever naming convention we end up deciding)?

TaiJuWu commented on kernel/msg_q.c:
```diff
@@ -187,6 +187,100 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
 	return result;
 }
 
+int z_impl_k_msgq_put_urgent(struct k_msgq *msgq, const void *data, k_timeout_t timeout)
```
From my old PR, a suggestion is `msg_prepend`

faxe1008 replied:
+1, urgent is kinda ambigous

TaiJuWu replied:
> I got a bit lost through the comments, why was you PR not approved? This feature is so simple, I don't get why it's not present yet.

Just that PR lacks reviewer and I did not  push it strongly.

TaiJuWu reviewed (commented):
Please add reference: https://github.com/zephyrproject-rtos/zephyr/issues/46339 to PR description.
Also this feature is good to be optional and it need more test.

alexpaschoaletto committed:
kernel: msgq: adding support to k_msgq_put_front

This commit introduces the k_msgq_put_front API for sending
messages to a queue in a LIFO scheme.

Signed-off-by: Alexander Paschoaletto <axelpinheiro@gmail.com>

alexpaschoaletto committed:
sample: msgq: adding a sample code for message queue base usage

this commit adds a sample code to illustrate the base usage
of message queues. a producer and a consumer thread work
together, exchanging messages in a FIFO (for normal payloads)
and LIFO (for higher priority payloads) schemes.

Signed-off-by: Alexander Paschoaletto <axelpinheiro@gmail.com>

TaiJuWu commented:
> > Please add reference: #46339 to PR description. Also this feature is good to be optional and it need more test.
> 
> done! not sure what you mean by "good to be optional", though. Proper tests will be done soon - I wanted some feedback on this contribution before taking the effort (if for some reason this was immediately discarded by reviewers, then it would be a waste of energy).

In my mind, the feature is not required for every developer so we can be optional and make the user to decide enabling it or not, this can reduce code size.

peter-mitsis reviewed (commented):
Aside from needing a proper test case, this looks to be on the right track--just some minor suggestions to tidy up a couple of corners.

alexpaschoaletto commented:
> Aside from needing a proper test case, this looks to be on the right track--just some minor suggestions to tidy up a couple of corners.

thank you for your time and comments, peter! I have applied most of your suggestions.

alexpaschoaletto commented on kernel/msg_q.c:
```diff
@@ -187,6 +187,96 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
+			msgq->read_ptr -= msgq->msg_size;
+			if (msgq->read_ptr < msgq->buffer_start) {
+				msgq->read_ptr = msgq->buffer_end - msgq->msg_size;
+			}
```
Clever one! I hadn't thought about that, thanks for highlighting the issue. Changes have been made.

alexpaschoaletto commented on kernel/msg_q.c:
```diff
@@ -187,6 +187,96 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
+			}
+
+			/* match read and write pointers */
+			msgq->write_ptr = msgq->read_ptr;
```
While writing about this contribution I came to a similar idea as yours, however I went a bit further in the search for optimization: since we have `msgq->read_ptr` already where we want to place the incoming message, it makes sense to just write directly to it instead of needing to manipulate `write_ptr` or, as you suggested, creating a temporary variable just for it.

Full explanation I left in the code:
```c
				/*
				 * writing a message to the head of the queue is
				 * less obvious, but it can be achieved by simply
				 * following these steps:
				 *
				 * 1. decrementing the read pointer. This effectively
				 * opens space for the incoming message in the head of
				 * the queue.
				 *
				 * 2. temporarily matching the write pointer with the
				 * read pointer. This way the message will be written
				 * in the recently opened space.
				 *
				 * 3. reverting the write pointer after the write to its
				 * original value. The read pointer, by its turn, should
				 * be always at the head, so it doesn't need reverting
				 * because we just wrote to the head.
				 *
				 * ...but this is inefficient because we need to go back
				 * and forth with the value of write_ptr; note that it
				 * becomes equal to read_ptr just to be restored afterwards.
				 * A much more efficient solution is to simply write the
				 * message directly to read_ptr's address after step 1,
				 * which avoids needing to perform steps 2 and 3 altogether.
				 */
				if (msgq->read_ptr == msgq->buffer_start) {
					msgq->read_ptr = msgq->buffer_end;
				}
				msgq->read_ptr -= msgq->msg_size;
				(void)memcpy(msgq->read_ptr, (char *)data, msgq->msg_size);
```

peter-mitsis commented:
> In my mind, the feature is not required for every developer so we can be optional and make the user to decide enabling it or not, this can reduce code size.

@TaiJuWu - I may be misunderstanding the intent behind your desire to make it optional for as I see it, it would in a sense automatically be so. That is, Zephyr uses garbage collection at build time to remove both unused functions and unused variables. Consequently, if no one uses the new urgent/front routine, it does not find its way into the final image.

peter-mitsis commented on kernel/msg_q.c:
```diff
@@ -150,13 +149,50 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
+				}
+			} else {
+				/*
+				 * writing a message to the head of the queue is
+				 * less obvious, but it can be achieved by simply
+				 * following these steps:
+				 *
+				 * 1. decrementing the read pointer. This effectively
+				 * opens space for the incoming message in the head of
+				 * the queue.
+				 *
+				 * 2. temporarily matching the write pointer with the
+				 * read pointer. This way the message will be written
+				 * in the recently opened space.
+				 *
+				 * 3. reverting the write pointer after the write to its
+				 * original value. The read pointer, by its turn, should
+				 * be always at the head, so it doesn't need reverting
+				 * because we just wrote to the head.
+				 *
+				 * ...but this is inefficient because we need to go back
+				 * and forth with the value of write_ptr; note that it
+				 * becomes equal to read_ptr just to be restored afterwards.
+				 * A much more efficient solution is to simply write the
+				 * message directly to read_ptr's address after step 1,
+				 * which avoids needing to perform steps 2 and 3 altogether.
```
Let's simplify this comment. Perhaps something like ...

				 * writing a message to the front of the queue is
				 * simple and effective: create space for the message
				 * by decrementing read_ptr, then copy the message
				 * to the newly created space

alexpaschoaletto replied:
alright, I have reduced the message. I still think the full rationale is valuable, though.

peter-mitsis commented on kernel/msg_q.c:
```diff
@@ -196,6 +243,15 @@ static inline int z_vrfy_k_msgq_put(struct k_msgq *msgq, const void *data,
+	K_OOPS(K_SYSCALL_MEMORY_READ(data, msgq->msg_size));
+
+	return z_impl_k_msgq_put_front(msgq, data, timeout);
+}
```
I think we are missing ...

#include <zephyr/syscalls/k_msgq_put_front_mrsh.c>

alexpaschoaletto replied:
really? ok, but what's that for?

peter-mitsis commented on include/zephyr/kernel.h:
```diff
@@ -4806,6 +4806,30 @@ int k_msgq_cleanup(struct k_msgq *msgq);
+ *
+ * This routine sends a message to the beginning (head) of message queue @a q.
+ * Messages sent with this method will be retrieved before any pre-existing
+ * messages in the queue.
```
Thanks for adding the note. I still think that it is important to have in there.

peter-mitsis commented on include/zephyr/kernel.h:
```diff
@@ -4785,7 +4785,7 @@ __syscall int k_msgq_alloc_init(struct k_msgq *msgq, size_t msg_size,
 
 /**
- * @brief Send a message to a message queue.
+ * @brief Send a message to the end of a message queue.
```
The commit message still has a referent to k_msgq_put_urgent instead of k_msgq_put_front.

TaiJuWu commented:
> > In my mind, the feature is not required for every developer so we can be optional and make the user to decide enabling it or not, this can reduce code size.
> 
> @TaiJuWu - I may be misunderstanding the intent behind your desire to make it optional for as I see it, it would in a sense automatically be so. That is, Zephyr uses garbage collection at build time to remove both unused functions and unused variables. Consequently, if no one uses the new urgent/front routine, it does not find its way into the final image.

Thanks for your explanation, I have not known this knowledge, I agree with you now.

peter-mitsis commented on kernel/msg_q.c:
```diff
@@ -124,11 +124,8 @@ int k_msgq_cleanup(struct k_msgq *msgq)
-int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout)
+static inline int put_msg_in_queue(struct k_msgq *msgq, const void *data, k_timeout_t timeout, bool put_at_back)
 {
-	__ASSERT(!arch_is_in_isr() || K_TIMEOUT_EQ(timeout, K_NO_WAIT), "");
```
Both checks are needed as blocking calls are not allowed at ISR level.

!arch_is_in_isr() : If true, then we are at at thread level, which is OK. Bail out of the assert.
Otherwise, it's false, (operating ISR level) and we must test for a blocking timeout. If K_NO_WAIT, it is non-blocking and this is OK from an ISR. Otherwise, it is a blocking timeout and the assert is triggered.

TaiJuWu replied:
I am fine for every decision you make ;)
It's purely my own bias, to be honest.

peter-mitsis commented:
Once the minor compliance check is fixed, I expect I will be giving this a +1.

alexpaschoaletto commented on samples/kernel/msg_queue/README.rst:
```diff
@@ -0,0 +1,50 @@
+
+   [producer] sending: 0
+   [producer] sending: 1
+   [producer] sending: A (urgent)
```
in the context of the sample, the publisher thread sends normal and urgent messages. It's not related to the API naming.

alexpaschoaletto committed:
tracing: msgq: adding dedicated tracing for k_msgq_put_front

This commit adds the tracing macros and functions related
specifically to the k_msgq_put_front API.

Signed-off-by: Alexander Paschoaletto <axelpinheiro@gmail.com>

JarmouniA reviewed (commented):
Mostly nits

TaiJuWu commented:
it is a test failure, is it excepted?

```
/**
 * @brief Put a message to a full queue for behavior test
 *
 * @details
 * - Thread A put message to a full message queue and then sleep
 * Thread B put a new message to the queue then pending on it.
 * - Thread A get all messages from message queue and check the behavior.
 *
 * @see k_msgq_put(), k_msgq_put_front()
 */
ZTEST(msgq_api_1cpu, test_msgq_pending)
{
	uint32_t rx_data;
	int pri = k_thread_priority_get(k_current_get()) - 1;
	int ret;

	k_msgq_init(&msgq1, tbuffer, MSG_SIZE, 2);
	ret = k_sem_init(&end_sema, 0, 1);
	zassert_equal(ret, 0);

	ret = IS_ENABLED(CONFIG_TEST_MSGQ_PREPEND) ?
		k_msgq_put_front(&msgq1, &data[1], K_NO_WAIT) :
		k_msgq_put(&msgq1, &data[0], K_NO_WAIT);
	zassert_equal(ret, 0);
	ret = IS_ENABLED(CONFIG_TEST_MSGQ_PREPEND) ?
		k_msgq_put(&msgq1, &data[0], K_NO_WAIT) :
		k_msgq_put_front(&msgq1, &data[1], K_NO_WAIT);
	zassert_equal(ret, 0);

	k_tid_t tid = k_thread_create(&tdata2, tstack2, STACK_SIZE,
					prepend_full_entry, &msgq1, NULL,
					NULL, pri, 0, K_NO_WAIT);

	/* that putting thread is being blocked now */
	k_sem_take(&end_sema, K_FOREVER);
	zassert_equal(tid->base.thread_state, _THREAD_PENDING);
	ret = k_msgq_get(&msgq1, &rx_data, K_FOREVER);
	zassert_equal(ret, 0);
	zassert_equal(rx_data, data[1]);

	ret = k_msgq_get(&msgq1, &rx_data, K_FOREVER);
	zassert_equal(ret, 0);
	zassert_equal(rx_data, msg3);

	ret = k_msgq_get(&msgq1, &rx_data, K_FOREVER);
	zassert_equal(ret, 0);
	zassert_equal(rx_data, data[0]);
	k_thread_abort(tid);
}
```

alexpaschoaletto commented on samples/kernel/msg_queue/sample.yaml:
```diff
@@ -0,0 +1,17 @@
+    integration_platforms:
+      - native_sim
+    tags:
+      - message_queue
```
ok, removed that.

alexpaschoaletto commented on kernel/msg_q.c:
```diff
@@ -150,13 +154,31 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
+				 * writing a message to the head of the queue can
+				 * be achieved by simply decrementing the read pointer
+				 * to open space at the front of the queue, then
+				 * copying the message to the newly created space.
```
same for this one, I've rewritten the comment.

alexpaschoaletto commented:
> it is a test failure, is it excepted?
> 
> The reason is described in [#92865 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/92865#discussion_r2208011326)

 I don't know if I understood. What exactly is to fail here?

andyross reviewed (approved):
Refresh +1

TaiJuWu commented:
> > it is a test failure, is it excepted?
> > The reason is described in [#92865 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/92865#discussion_r2208011326)
> 
> I don't know if I understood. What exactly is to fail here?

This test covers the behavior of putting a message into the message queue.
In theory, it should pass, but in reality, this test fails.
The reason is that when the buffer is full, Thread A gets pended (blocked).
If Thread B later calls the get function, it will wake up Thread A, allowing it to resume and put the message into the queue.
In this situation, we need to know whether Thread A should continue with put to front or put to end.

Does this scenario make sense?
